### DISCUSSION TITLE
Incorrect relation with aliases

### DISCUSSION THREAD

**Discussion by leonardomarcao** [sqlalchemy/sqlalchemy#9262] [Usage Questions]
I have a main entity called opportunity and 2 other entities called contact_media and company_media that reference opportunity via 2 foreign keys for each. I also have a polymorphic entity called media which has a child entity called media_type. The entities contact_media and company_media are also linked to the media entity, as described below.

contact_media.contact_id <- opportunity.contact_id
company_media.customer_id <- opportunity.customer_id
media.id <- contact_media.id
media.id <- company_media.id
media_type.id <- media.type_id

The problem.
When I go to relate these entities dynamically and using aliased, at the time of relating these entities the reference target gets lost, i.e., instead of the media (anon_1) that should link to the contact_media (anon_2), it ends up linking to the company_media, then bringing the incorrect media_type and the correct media. 

Since it is something dynamic, I can't reproduce such an "error", only provide the code snippet that generates "this stuff" dynamically and automatically. The models themselves are okay. 

Could someone help me?

I don't have any way to reproduce these steps, only my code. Sorry for that.

**code snippet**

```python
	@classmethod
	def __includeRelationships(cls, request, baseQuery):
		includes = []
		requestIncludes = request.query_params.get('includes')
		if requestIncludes is not None:
			includes = requestIncludes.split(",")
			allJoins = []
			joinedAttributes = {}
			for include in includes:
				cascadeEagerLoadTrain = None
				cascadeJoinTrain = baseQuery
				cascadeRawParent = cascadeParent = cls

				attributeName = None
				for cascade in include.split("."):
					if attributeName is None:
						attributeName = cascade
					else:
						attributeName += "." + cascade

					if hasattr(cascadeParent, cascade):
						cascadeAttribute = getattr(cascadeRawParent, cascade)
						cascadeRawParent = cascadeAttribute.property.mapper.class_

						if attributeName not in joinedAttributes:
							cascadeParent = aliased(cascadeRawParent)
							joinedAttributes[attributeName] = cascadeParent

							if hasattr(cascadeParent, 'ACLApplies') and cascadeParent.ACLApplies(request) and not (hasattr(cls, "__BYPASSS_CHILD_ACL__") and cls.__BYPASSS_CHILD_ACL__):
								enforcingFilter = cascadeParent.__enforceIncludesAccessControl(request)
								if enforcingFilter is not None:
									cascadeJoinTrain = cascadeJoinTrain.join(cascadeParent, cascadeAttribute.and_(enforcingFilter), isouter=True)
								else:
									cascadeJoinTrain = cascadeJoinTrain.join(cascadeParent, cascadeAttribute, isouter=True)
							else:
								cascadeJoinTrain = cascadeJoinTrain.join(cascadeParent, cascadeAttribute, isouter=True)

							cascadeJoinTrain = cls.__applyFilterOnRelationship(request, cascadeJoinTrain, attributeName, cascadeParent)
							cascadeJoinTrain = cls.__applyBooleanExpressionIncludingRelashionship(request, cascadeJoinTrain, attributeName, cascadeParent)
						else:
							cascadeParent = joinedAttributes[attributeName]
						if cascadeEagerLoadTrain is None:
							cascadeEagerLoadTrain = contains_eager(cascadeAttribute, alias=cascadeParent)
						else:
							cascadeEagerLoadTrain = cascadeEagerLoadTrain.contains_eager(cascadeAttribute, alias=cascadeParent)
					else:
						break

				if cascadeEagerLoadTrain:
						allJoins.append(cascadeEagerLoadTrain)

				baseQuery = cascadeJoinTrain
			if allJoins:
				baseQuery = baseQuery.options(*allJoins)
		return (baseQuery, includes)
```

**Comment by zzzeek** [comment]
hi -

as I was telling someone [yesterday](https://github.com/sqlalchemy/sqlalchemy/discussions/9250#discussioncomment-4885097), contains_eager() is very difficult to use for highly nested join structures, and to do it dynamically as you seem to be doing above would be enormously difficult, and potentially not possible, as contains_eager() likely still has some edge cases that don't produce the preferred entries, especially with table inheritance models.

To work with code like the above you would need to start with something simple and apply very heavy test driven development to the results, writing out full tests for as many scenarios as possible.  That would allow you to isolate the specific change / capability that makes it start to fail, which would then give you a starting point as to how to illustrate a self-contained case here that we could decide why it works the way it does, or if there's a shortcoming in contains_eager().

Sorry I can't give you more to go on than that, what you are doing above is way way out there and not something I'd ever try to get working.

**Reply by leonardomarcao** [reply]
@zeeeeeb 

`LEFT OUTER JOIN media_type AS media_type_2 ON media_type_2.id = anon_1.media_type_id` the join condition should be with anon_3 alias instead anon_1. This happens exactly on the join `cascadeJoinTrain = cascadeJoinTrain.join(cascadeParent, cascadeAttribute, isouter=True)`

thank you for your reply, but after this, do you have any idea?

**Reply by zzzeek** [reply]
when I get a query like this, or really one that's about 1/4 the size of this, I work from the full model and reproducing code and then have to spend many hours usually understanding what's going on, which usually ends up being an issue where SQLAlchemy sees the same entity mentioned in two places in the query, and is not able to correctly figure out how that entity should be "aliased" based on its position in the query.   This can occur when the same entity is used multiple times in an ambiguous way but also is where the majority of bugs with contains_eager() occur in any case.

So I can say that yes, an issue like "the LEFT OUTER JOIN should be to anon_3 and not anon_1" is extremely common, but there's not a switch that can just "fix" it, it has to do with how the query was constructed and the very specific details of the model and how it interacts.

I will reiterate that it looks like your goal is to receive a string like "model.linka.linkb.linkc, ..." and parse it.  I would **not do that**.  Write SQLAlchemy queries directly instead without having to invent a complex translation layer.

**Reply by leonardomarcao** [reply]
Thank you so much @zzzeek.

**Reply by CaselIT** [reply]
> I will reiterate that it looks like your goal is to receive a string like "model.linka.linkb.linkc, ..." and parse it. I would **not do that**. Write SQLAlchemy queries directly instead without having to invent a complex translation layer.

I do have something like this at work, but it translates it in loaders to provide to the `options()`, it does not try to manually construct the query itself and then use contains_egaer. Basically it's a combination of selectinload and load_only possibly nested over multiple levels.

So far no bug where reported on it and it has been remarkably easy to maintain over multiple sqlalchemy versions (was originally written on 1.3 using query, and other than moving it to select in 1.4 it had no change)

Sadly it's not public

**Comment by leonardomarcao** [comment]
Hi there. 

I'm back to say that I fixed this through of workaround which makes an explicit join as @zzzeek said.
So, my class receives an attribute called __explicit_join__ and I use this - if that is the case - to specify the clause of join. 
I keep dynamic queries and specify the join - if necessary - case of the model class containing the attribute. 

Is very interesting to keep trying to make a complex translation layer because this abstract all things and saves a lot of time!

Thank you!
